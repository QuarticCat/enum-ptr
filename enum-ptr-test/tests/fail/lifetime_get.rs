#![allow(dead_code)]

use enum_ptr::{get_mut, get_ref, Compact, EnumPtr};

#[derive(EnumPtr)]
#[repr(C, usize)]
enum Foo<'a, 'b> {
    A(Option<&'a mut i32>),
    B(Option<&'b mut i32>),
}

fn test_get_ref() {
    let foo: Compact<_> = Foo::A(None).into();
    let foo_ref = get_ref!(foo, Foo::A);
    drop(foo);
    drop(foo_ref);
}

fn test_get_mut() {
    let mut foo: Compact<_> = Foo::A(None).into();
    let foo_mut = get_mut!(foo, Foo::A);
    drop(foo);
    drop(foo_mut);
}

fn main() {}
