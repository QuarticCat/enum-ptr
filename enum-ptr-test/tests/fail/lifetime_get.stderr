error[E0505]: cannot move out of `foo` because it is borrowed
  --> tests/fail/lifetime_get.rs:15:10
   |
13 |     let foo: Compact<_> = Foo::A(None).into();
   |         --- binding `foo` declared here
14 |     let foo_ref = get_ref!(&foo, Foo::A);
   |                            ---- borrow of `foo` occurs here
15 |     drop(foo);
   |          ^^^ move out of `foo` occurs here
16 |     drop(foo_ref);
   |          ------- borrow later used here

error[E0505]: cannot move out of `foo` because it is borrowed
  --> tests/fail/lifetime_get.rs:22:10
   |
20 |     let mut foo: Compact<_> = Foo::A(None).into();
   |         ------- binding `foo` declared here
21 |     let foo_mut = get_mut!(&mut foo, Foo::A);
   |                            -------- borrow of `foo` occurs here
22 |     drop(foo);
   |          ^^^ move out of `foo` occurs here
23 |     drop(foo_mut);
   |          ------- borrow later used here
